src/hotspot/share/opto/memnode.cpp

```diff
@@ -2761,9 +2761,441 @@ Node *StoreNode::Ideal(PhaseGVN *phase, bool can_reshape) {
 }
 
+// Link together multiple stores (B/S/C/I) into a longer one.
+Node* StoreNode::Ideal_merge_stores(PhaseGVN* phase) {
+  int opc = Opcode();
+  if (opc != Op_StoreB && opc != Op_StoreC && opc != Op_StoreI) {
+    return nullptr;
+  }
+
+  // If we can merge with use, then we must process use first.
+  StoreNode* use = can_merge_with_use(phase, true);
+  if (use != nullptr) {
+    return nullptr;
+  }
+
+  // Check if we can merge with at least one def.
+  StoreNode* def = can_merge_with_def(phase, true);
+  if (def == nullptr) {
+    return nullptr;
+  }
+
+  // Now we know we can merge at least two stores.
+  ResourceMark rm;
+  Node_List merge_list;
+  merge_list.push(this);
+
+  uint merge_list_max_size = 8 / memory_size();
+  assert(merge_list_max_size >= 2 &&
+         merge_list_max_size <= 8 &&
+         is_power_of_2(merge_list_max_size),
+         "must be 2, 4 or 8");
+
+  // Collect list of stores
+  while (def != nullptr && merge_list.size() <= merge_list_max_size) {
+    merge_list.push(def);
+    def = def->can_merge_with_def(phase, true);
+  }
+
+  int pow2size = round_down_power_of_2(merge_list.size());
+  assert(pow2size >= 2, "must be merging at least 2 stores");
+
+  // Create / find new value:
+  Node* new_value = nullptr;
+  int new_memory_size = memory_size() * pow2size;
+  if (in(MemNode::ValueIn)->Opcode() == Op_ConI) {
+    // Collect all constants
+    jlong con = 0;
+    jlong bits_per_store = memory_size() * 8;
+    jlong mask = (((jlong)1) << bits_per_store) - 1;
+    for (int i = 0; i < pow2size; i++) {
+      jlong con_i = merge_list.at(i)->in(MemNode::ValueIn)->get_int();
+      con = con << bits_per_store;
+      con = con | (mask & con_i);
+    }
+    new_value = phase->longcon(con);
+  } else {
+    Node* first = merge_list.at(pow2size-1);
+    new_value = first->in(MemNode::ValueIn);
+    Node* base_last;
+    jint shift_last;
+    bool is_true = is_con_RShift(in(MemNode::ValueIn), base_last, shift_last);
+    assert(is_true, "must detect con RShift");
+    if (new_value != base_last && new_value->Opcode() == Op_ConvL2I) {
+      // look through
+      new_value = new_value->in(1);
+    }
+    if (new_value != base_last) {
+      // new_value is not the base
+      return nullptr;
+    }
+  }
+
+  if (phase->type(new_value)->isa_long() != nullptr && new_memory_size <= 4) {
+    new_value = phase->transform(new ConvL2INode(new_value));
+  }
+
+  assert((phase->type(new_value)->isa_int() != nullptr && new_memory_size <= 4) ||
+         (phase->type(new_value)->isa_long() != nullptr && new_memory_size == 8),
+         "new_value is either int or long, and new_memory_size is small enough");
+
+  Node* first = merge_list.at(pow2size-1);
+  Node* new_ctrl = in(MemNode::Control); // must take last: after all RangeChecks
+  Node* new_mem  = first->in(MemNode::Memory);
+  Node* new_adr  = first->in(MemNode::Address);
+  const TypePtr* atp = TypeRawPtr::BOTTOM;
```

Comment:
What does C2 prevent from moving a load from one of the array elements to above the raw store, assuming they are independent?

Reply:
I don't know. I have added a test now for that, but probably need more tests.
@vnkozlov @rwestrel  do you think I need to do something about this?

Reply:
Yes, this could be the issue if we don't check that memory accesses offset overlaps.

There was issue we recently fixed [#16015](https://github.com/openjdk/jdk/pull/16015) where we did not take into account wider memory access to array.

Reply:
@vnkozlov Hmm, ok. You seem to say that it is mostly a question of bugs that exist elsewhere. But my change here would not really change things then, since we could always trigger those things with Unsafe anyway, right?

Or is there something intrinsically wrong about using **raw store**?

Is there any alternative for me here? @TobiHartmann suggested I could use CPU barriers.

Reply:
Adding `MemBarCPUOrder` after this store is good suggestion. But hte need for it depends how you replace all merged stores with new wide store and attach following loads. Looking on changes and it seems you allow this optimization for sequential fields of normal object and not just array (I don't see check for AryPtrType). Such accesses will have separate memory slices (aliases). If you attach all memory edges of following related Load nodes to new wide Store we may be fine because loads will not pass the store. If not, we have a problem. It needs to be verified. Also looking on `find_previous_store()` and `can_see_stored_value()` they seem have checks which may prevent Loads from looking above such Store.

We need more testing for such cases. I see only 2 test methods which do that: test300R and test300a.
Also all since all init values the same it is difficult to see if Load node get correct value.

Reply:
Right, for sequential fields it's definitely a problem but I thought this optimization is limited to arrays? I hit this with a similar optimization in Valhalla and as a quick workaround, I'm using the `C2_UNSAFE_ACCESS` decorator for the accesses which will emit a membar via `C2Access::needs_cpu_membar()`. It's worth checking though if it is really needed.